src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,690 @@
+        /// and replace its m_WriteList arraylist field.
+        /// </summary>
+        private sealed class ConnectionArrayList : ArrayListWrapper
+        {
+            public ConnectionArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                if (s_connectionType.IsInstanceOfType(value))
+                {
+                    // Replace the HttpWebRequest arraylist inside this Connection object,
+                    // which allows us to intercept each new HttpWebRequest object added under
+                    // this Connection.
+                    ArrayList originalArrayList = s_writeListField.GetValue(value) as ArrayList;
+                    HttpWebRequestArrayList newArrayList = new HttpWebRequestArrayList(originalArrayList ?? new ArrayList());
+
+                    s_writeListField.SetValue(value, newArrayList);
+                }
+
+                return base.Add(value);
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for Connection.m_WriteList. The goal here is to
+        /// intercept all new HttpWebRequest objects being added to Connection.m_WriteList
+        /// and notify the listener about the HttpWebRequest that's about to send a request.
+        /// It also intercepts all HttpWebRequest objects that are about to get removed from
+        /// Connection.m_WriteList as they have completed the request.
+        /// </summary>
+        private sealed class HttpWebRequestArrayList : ArrayListWrapper
+        {
+            public HttpWebRequestArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                HttpWebRequest request = value as HttpWebRequest;
+                if (request != null)
+                {
+                    s_instance.RaiseRequestEvent(request);
+                }
+
+                return base.Add(value);
+            }
+
+            public override void RemoveAt(int index)
+            {
+                HttpWebRequest request = base[index] as HttpWebRequest;
+                if (request != null)
+                {
+                    HttpWebResponse response = s_httpResponseAccessor(request);
+                    if (response != null)
+                    {
+                        s_instance.RaiseResponseEvent(request, response);
+                    }
+                }
+
+                base.RemoveAt(index);
+            }
+        }
+
+        #endregion
+
+        #region private methods
+
+        /// <summary>
+        /// Private constructor. This class implements a singleton pattern and only this class is allowed to create an instance.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        private void RaiseRequestEvent(HttpWebRequest request)
```

Comment:
I agree that having the same logic here and in HttpClient is a good thing. De-duplication could be based on the presence of headers in the request.
Events names then should be renamed to be consistent with HttpClient, the whole IsEnabled flow also goes here.

The problem is that the change to AppInsights was done in  a last minute before the code freeze and now it is probably too late to do any changes there.

I see following options:
**1**. Do the right thing here, in DiagnosticSource, make it fully repeat HttpClient instrumentation on corefx: create activity and inject headers. 
It requires some minor changes in AppInsights @brahmnes, we need to check if AI team is fine allowing us to do this as a bugfix. If they are fine, I can finish these changes and fix AI.

If we don't do it now, we either 
-  leave current behavior (no activities, consumer injects headers) **forever** 
- OR have to deal with events deprecation later (e.g. in the next major DiagnosticSource version)

**2**. Do not have this hook in DiagnosticSource for now, keep it in AppInsights (it is copy-pasted for testing). Add the hook in the next major release of DiagnosticSource, make some future AppInsights release compatible with it and remove hook from AppInsights.
However there is an agreement with AI team that the the hook will be moved to DiagnosticSource.

**3**. Leave current behavior, but as i mentioned before we either have to live with this forever or deal with deprecation problems later.

I strongly prefer option 1 as long as AppInsights team agree on taking fix for this release.

Otherwise, option 2 is not that bad: only AI knows how to properly use the hook and it is custom for it, so it lives in AI codebase.

@brahmnes could you discuss this options with AI team?